Implementing gain from YAML files for calibration

IzaakWN opened the pull request:
Implementing gain from YAML files for calibration

IzaakWN committed:
add options

IzaakWN committed:
Merge branch 'hgcal-condformat-HGCalNANO-13_2_0_pre2' of github.com:CMS-HGCAL/cmssw into hgcal-condformat-HGCalNANO-13_2_0_pre2

IzaakWN committed:
implement gain

jniedzie reviewed (commented):
Hi @IzaakWN, thanks for the PR, the HGCalRecHitProducer part looks good to me! I let others have a look at other parts and merge it.

IzaakWN commented:
Hi @jniedzie, thanks for the feedback. I still have to validate it and iterate with @pfs, so I left it as a draft for now.

IzaakWN committed:
fix bugs; validate

IzaakWN committed:
clean

IzaakWN commented:
Hi @pfs, @yulunmiao,

I implemented the loading of the gain from [`HGCalConfigESSourceFromYAML.cc`](https://github.com/CMS-HGCAL/cmssw/blob/e7c0c2ce5bb2ebcac4fb48e42d976116bf693c58/CalibCalorimetry/HGCalPlugins/plugins/HGCalConfigESSourceFromYAML.cc#L113-L118) into [`HGCalCalibrationParameterProvider`](https://github.com/CMS-HGCAL/cmssw/blob/fb901c02e2453e1240d129ea789eb82796d71441/RecoLocalCalo/HGCalRecAlgos/interface/HGCalCalibrationParameterProvider.h#L16) in [`HGCalRecHitProducer.cc`](https://github.com/CMS-HGCAL/cmssw/blob/e7c0c2ce5bb2ebcac4fb48e42d976116bf693c58/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitProducer.cc#L113-L136). As the final YAML structure is still not clear, I have implemented placeholders, assuming 3 ROCs, i.e. 6 half-ROCs:
https://github.com/CMS-HGCAL/cmssw/blob/e7c0c2ce5bb2ebcac4fb48e42d976116bf693c58/CalibCalorimetry/HGCalPlugins/plugins/HGCalConfigESSourceFromYAML.cc#L113-L118
where `gain = 1` is used in `charge = ( ADC - pedestal ) / 2^10 * gain * 80 fC` as a first approximation: https://github.com/CMS-HGCAL/cmssw/blob/fb901c02e2453e1240d129ea789eb82796d71441/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationAlgorithms.dev.cc#L50-L62

This PR can already be merged on my part. However, there are still two points that need to be discussed for further development:
1. The pedestals, common mode, and gain stored in `HGCalCalibrationParameterProvider` are actually still not propagated to `HGCalRecHitCalibrationAlgorithms.dev.cc`. There are still placeholders used: https://github.com/CMS-HGCAL/cmssw/blob/fb901c02e2453e1240d129ea789eb82796d71441/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationAlgorithms.dev.cc#L116-L129 @yulunmiao, can you help with this, because I am not sure how to do this correctly with Alpaka?
2. Of second order: The gain is defined per ROC, but right now, I added it to `HGCalCalibrationParameterProvider` for every channel. This can probably be optimized, but I am not sure how to do this with dense indexing, which @yulunmiao implemented during the Patatrack Hackathon.

Thanks!
Izaak

pfs commented on CalibCalorimetry/HGCalPlugins/plugins/HGCalConfigESSourceFromYAML.cc:
```diff
@@ -88,33 +89,60 @@ class HGCalConfigESSourceFromYAML : public edm::ESProducer, public edm::EventSet
-      //for (const auto& params : yaml_file) { // loop through nodes
-      //  std::cout << "HGCalConfigESSourceFromYAML::parseYAML: Found key " << params.first.as<std::string>() << std::endl;
+      // PARSE META DATA NODE: place holders
+      cond->moduleConfigs[0].charMode = (bool) 0;
```
this and the next lines will break the test runs we have been using from 2022
on one hand we should implement  it as what the final YAML files will contain
maybe one could have a "version" config parameter. In case we configure as "version=2022" it parses the 2022 version of the YAMLs?

IzaakWN replied:
Yes, I can maybe add this version parameter to manually set it in the python config until we implemented taking it from full YAML so it's easier to run `tb_raw2reco.py` without changing this hardcoded placeholder.

IzaakWN replied:
@pfs, is it okay with manual overrides like this: https://github.com/CMS-HGCAL/cmssw/pull/31/commits/6648419545faa0229485297b542223e2ecc2257c
This would allow us to run the whole setup without relying on the current state of the YAML parser, as we haven't converged on a YAML format yet, let alone parsing it for relays?

pfs replied:
+1 (i imagine the stress)

pfs commented on RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationAlgorithms.dev.cc:
```diff
@@ -20,47 +20,61 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
       for (auto index : elements_with_stride(acc, digis.metadata().size())) {
         recHits[index].detid() = static_cast<uint32_t>(digis[index].electronicsId());
-        recHits[index].energy() = ADC_to_energy(digis[index].adc(),digis[index].tot(),digis[index].tctp());
+        recHits[index].energy() = ADC_to_float(recHits[index].energy(),digis[index].tot(),digis[index].tctp());
```
as this is the first conversion one should use
ADC_to_energy(digis[index].adc(),digis[index].tot(),digis[index].tctp());

IzaakWN replied:
Ah yes, this changed after switching the order...

pfs commented on RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationAlgorithms.dev.cc:
```diff
@@ -20,47 +20,61 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
-      auto ADC_to_energy = [&](uint32_t ADC,uint32_t TOT, uint8_t tctp) { return float(tctp>0 ? TOT : ADC); };
-
+      auto ADC_to_float = [&](float ADC, uint32_t TOT, uint8_t tctp) { return float(tctp>0 ? TOT : ADC); };
+      
```
as this is the first conversion one should use
 auto ADC_to_energy = [&](uint32_t ADC,uint32_t TOT, uint8_t tctp) { return float(tctp>0 ? TOT : ADC); };

pfs commented:
Thanks @IzaakWN i left some comments in the code. What you write above seems reasonable.
I guess now is a matter of trying and checking if before/after each step things make sense.
The sequence should be 
DIGI -> Float -> Pedestal -> CM subtraction -> Charge

IzaakWN commented:
Hi @pfs, I can't see your review comments. You probably need to submit them collectively before I can see them.

In the mean time @ywkao has propagated the calibration parameters to the kernels.

pfs commented:
@IzaakWN  Ok. If you click on "Files Changed" the comments should be interleaved with the code, no?
Anyway thank to you and @ywkao. 
Notice that the output of the DQM module has some extra columns. The ESProducer needs to be adapted to read (and discard) these extra columns, otherwise you may get strange values.

| id | pedestal | noise (sigma) | slope (cm) | offset (CM) | slope (BX-1) | offset (BX-1) |

https://github.com/CMS-HGCAL/cmssw/blob/hgcal-condformat-HGCalNANO-13_2_0_pre2/DQM/HGCal/plugins/HGCalDigisClientHarvester.cc#L244-L249

pfs commented:
> Hi @pfs, I can't see your review comments. You probably need to submit them collectively before I can see them.

ah sorry i was not aware of the "submit" thingy

IzaakWN commented:
> ah sorry i was not aware of the "submit" thingy

No problem, happened to me several times... I am sure many comments are missed in this way...

I will implement your suggestions and merge with @ywkao's changes before the beam test hopefully.

IzaakWN committed:
float -> uint32_t; recHits[index].energy() -> digis[index].adc()

IzaakWN committed:
add manual override for charMode and gain in YAML files (for quick testing)

IzaakWN committed:
implement Pedro's suggestions; char/uint32_t -> uint8_t for gain and injgain

IzaakWN committed:
Merge branch 'hgcal-condformat-HGCalNANO-13_2_0_pre2' into hgcal-condformat-HGCalNANO-13_2_0_pre2_gain

IzaakWN committed:
remove space

pfs commented:
Can this PR be rebased to the latest version? Right now it has conflicts. Thanks

IzaakWN commented:
> Can this PR be rebased to the latest version? Right now it has conflicts. Thanks

Yes, I was working on resolving previous conflicts, but new ones arose and the structure of the conditions provider has changed quite a bit.

IzaakWN committed:
Merge branch 'hgcal-condformat-HGCalNANO-13_2_0_pre2' into hgcal-condformat-HGCalNANO-13_2_0_pre2_gain

IzaakWN committed:
Merge branch 'hgcal-condformat-HGCalNANO-13_2_0_pre2' into hgcal-condformat-HGCalNANO-13_2_0_pre2_gain

IzaakWN committed:
add Alpaka ESProducers for the gain (still in development)

IzaakWN commented:
Hi all,

Just to give a status before I will be offline for two weeks: @ywkao recently pushed a working example of an Alpaka ESProducer to provide the channel-level calibration parameters like the pedestals: https://github.com/CMS-HGCAL/cmssw/commit/3f87f2217765518dfcb5386289e5a6e108ec51cc

With this example, I have added a second ROC-level SoA and a second Alpaka ESProducer to provide ROC-level configuration parameters like the gain for charge conversion during the calibration step. This is still under development.

We are discussing with Andrea Bocci & Matti Kortelainen if we could instead have one ESProducer that can provide both the channel-level (filled by text files) and ROC-level SoA (filled by configuration parameter values from YAML). Unfortunately `edm::ESProducts` does not seem  to work for Alpaka ESProducers at the moment, so we will try probably try to create two separate production functions.

Regarding the parsing of the YALM files: This is something I want to do later, once we have an agreed upon format/structure. Until then I propose to implement parameters that allow the user to manually set/override the gain, characterization mode, etc. without changing hardcoded placeholders and recompiling each time.

Cheers,
Izaak

ywkao committed:
Fix a mismatched module name

ywkao committed:
Swap order of CM subtraction and charge conversion

IzaakWN committed:
Merge branch 'hgcal-condformat-HGCalNANO-13_2_0_pre2' into hgcal-condformat-HGCalNANO-13_2_0_pre2_gain

IzaakWN committed:
make HGCalConfigurationESProducer dependend on HGCalCondSerializableModuleInfoRcd; move ESProducers to hgcalrechit namespace; ...

IzaakWN committed:
implement DependentRecordImplementation for HGCalCondSerializableConfigRcd; rename ESInputTags

IzaakWN committed:
propagate gain from YAML via HGCalConfigurationESProducer

IzaakWN committed:
add 32 bit masks to ElectronicsID; clean print out

IzaakWN committed:
clean code

IzaakWN committed:
fix bug in cpi.getSize(); clean code

IzaakWN committed:
remove debugging print out

IzaakWN commented:
This PR should be ready for merging.

More development is needed, but the current implementation works as expected for a given gain. Many small changes were made, so if we already merge we avoid the need for resolving more potential conflicts.

Some notes as summary of changes below.

## YAMLs configuration files
1. The full parser of YAML configs will come later, as the file structure is not settled yet. For now the configuration is filled with placeholders for `charMode` and `gain`, which can be manually overridden via the python configuration.
2. In the future we might replace the ESProducer [`HGCalConfigESSourceFromYAML`](https://github.com/CMS-HGCAL/cmssw/blob/hgcal-condformat-HGCalNANO-13_2_0_pre2_gain/CalibCalorimetry/HGCalPlugins/plugins/HGCalConfigESSourceFromYAML.cc) that reads the YAML files with a standalone class.

## Configuration parameter Alpaka SoA & ESProducers
3. In addition to the SoA and ESProducer ([`HGCalCalibrationESProducer`](https://github.com/CMS-HGCAL/cmssw/blob/hgcal-condformat-HGCalNANO-13_2_0_pre2_gain/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationESProducer.cc)) for "channel-level" calibration parameters (like pedestals), a "ROC-level SoA" and Alpaka ESProducer ([`HGCalConfigurationESProducer`](https://github.com/CMS-HGCAL/cmssw/blob/hgcal-condformat-HGCalNANO-13_2_0_pre2_gain/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitConfigurationESProducer.cc)) is implemented to propagate the configuration parameters (for now only gain) from the YAML ESProducer.
https://github.com/CMS-HGCAL/cmssw/blob/0593a973b1181a79ab1a56f260f7485e593c80a0/RecoLocalCalo/HGCalRecAlgos/interface/HGCalCalibrationParameterSoA.h#L33-L37
4. In the future, we may add a third SoA and ESProducer to propagate channel-level configuration parameters.
5. To get the ESProducer working, the `HGCalCondSerializableConfigRcd` record had to be made dependent on `HGCalCondSerializableModuleInfoRcd` to retrieve the module info, which is needed to set the size for of the dense index map in `HGCalConfigurationESProducer`:
https://github.com/CMS-HGCAL/cmssw/blob/c2eac662ab9874eadbcc385b841df4a18885e539/CondFormats/DataRecord/interface/HGCalCondSerializableConfigRcd.h#L25-L27 https://github.com/CMS-HGCAL/cmssw/blob/c2eac662ab9874eadbcc385b841df4a18885e539/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitConfigurationESProducer.cc#L55-L64
6. A future TODO: implement some assertions/checks of the value ranges in the SoA cells and that the SoA dimensions really match the module being calibrated.

## Calibration
7. Charge conversion happens after pedestal and common mode subtraction:
https://github.com/CMS-HGCAL/cmssw/blob/c2eac662ab9874eadbcc385b841df4a18885e539/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationAlgorithms.dev.cc#L129-L140
8. Charge conversion looks like:
https://github.com/CMS-HGCAL/cmssw/blob/c2eac662ab9874eadbcc385b841df4a18885e539/RecoLocalCalo/HGCalRecAlgos/plugins/alpaka/HGCalRecHitCalibrationAlgorithms.dev.cc#L52-L66

## Other changes
9. To get the 32-bit electronicsId for the modules, I implemented a new `enum` set of masks in the `HGCalElectronicsId` class:
https://github.com/CMS-HGCAL/cmssw/blob/c2eac662ab9874eadbcc385b841df4a18885e539/DataFormats/HGCalDigi/interface/HGCalElectronicsId.h#L39-L46 https://github.com/CMS-HGCAL/cmssw/blob/c2eac662ab9874eadbcc385b841df4a18885e539/DataFormats/HGCalDigi/src/HGCalElectronicsId.cc#L46-L50 https://github.com/CMS-HGCAL/cmssw/blob/c2eac662ab9874eadbcc385b841df4a18885e539/EventFilter/HGCalRawToDigi/plugins/HGCalRawToDigi.cc#L177-L179

## Validation
10. Validated as:
```
cd ~/hgcal/
source $VO_CMS_SW_DIR/cmsset_default.sh
export SCRAM_ARCH=el8_amd64_gcc11 # lxplus8
cmsrel CMSSW_13_2_0_pre2
cd CMSSW_13_2_0_pre2/src
cmsenv
git cms-checkout-topic -u CMS-HGCAL:hgcal-condformat-HGCalNANO-13_2_0_pre2
USER_CXXFLAGS="-g -D=EDM_ML_DEBUG" scram b -j8

cmsRun $CMSSW_BASE/src/EventFilter/HGCalRawToDigi/test/tb_raw2reco.py activeECONDs=0 fedId=0 maxEvents=1000 cbHeaderMarker=0x0 gain=1 | head -n 1000
```

Output should look something like the table below. I validated the charge conversion, and that it scales as expected for different given values of the gain.
```
HGCalRecHitCalibrationAlgorithms: Input recHits:
i       detid   energy          time            flags
0	0	0.000000	0.000000	0
1	1	0.000000	0.000000	0
2	2	0.000000	0.000000	0
3	3	0.000000	0.000000	0
4	4	0.000000	0.000000	0
5	5	0.000000	0.000000	0
6	6	0.000000	0.000000	0
7	7	0.000000	0.000000	0
8	8	0.000000	0.000000	0
9	9	0.000000	0.000000	0
HGCalRecHitCalibrationAlgorithms: RecHits after pedestal calibration: 
0	0	0.000000	0.000000	0
1	1	0.000000	0.000000	0
2	2	0.000000	0.000000	0
3	3	0.000000	0.000000	0
4	4	0.000000	0.000000	0
5	5	0.000000	0.000000	0
6	6	0.000000	0.000000	0
7	7	0.000000	0.000000	0
8	8	0.000000	0.000000	0
9	9	0.000000	0.000000	0
HGCalRecHitCalibrationAlgorithms: Digis after CM calibration: 
0	0	-65.431450	0.000000	0
1	1	-61.946053	0.000000	0
2	2	-73.598755	0.000000	0
3	3	-75.142593	0.000000	0
4	4	-83.182808	0.000000	0
5	5	-84.483444	0.000000	0
6	6	-83.699867	0.000000	0
7	7	-85.003685	0.000000	0
8	8	-67.012497	0.000000	0
9	9	-70.999428	0.000000	0
HGCalRecHitCalibrationAlgorithms: RecHits after charge conversion: 
0	0	-1675.045117	0.000000	0
1	1	-1585.818945	0.000000	0
2	2	-1884.128125	0.000000	0
3	3	-1923.650391	0.000000	0
4	4	-2129.479883	0.000000	0
5	5	-2162.776172	0.000000	0
6	6	-2142.716602	0.000000	0
7	7	-2176.094336	0.000000	0
8	8	-1715.519922	0.000000	0
9	9	-1817.585352	0.000000	0
```

pfs commented:
Note: we should add mabe rebase the pre3 branch to include this in the new baseline.
